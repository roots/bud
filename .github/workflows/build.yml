name: build

on:
  push:
    paths:
      - packages/**/*/package.json
      - packages/**/*/src/*
      - packages/**/*/src/**/*
      - yarn.lock
    pull_request:
      branches:
        - stable

jobs:
  build:
    strategy:
      matrix:
        platform:
          - ubuntu-latest
        node:
          - 15.11.0
    name: test ${{matrix.node}}/${{matrix.platform}}
    runs-on: ${{matrix.platform}}
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - uses: actions/setup-node@v1
        with:
          node-version: ${{matrix.node}}
      - run: yarn --immutable
      - run: yarn build:ci
      - run: yarn pkg
