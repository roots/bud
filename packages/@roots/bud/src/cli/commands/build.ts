import * as Oclif from '@oclif/command'

import {Command} from '../Command'
import * as flags from '../flags'

/**
 * @public
 */
export default class Build extends Command {
  /**
   * @public
   */
  public static id = 'build'

  /**
   * @public
   */
  public static title = 'build'

  /**
   * @public
   */
  public static description = 'compile source assets'

  /**
   * @public
   */
  public static examples = [`$ bud build`, `$ bud build --cache`]

  /**
   * @public
   */
  public static flags = {
    ...Command.flags,
    ...flags.target,
    ...flags.location,

    mode: Oclif.flags.string({
      description: 'compiler mode',
      default: 'production',
      options: ['development', 'production'],
      hidden: true,
    }),

    cache: Oclif.flags.boolean({
      allowNo: true,
      default: true,
      description: 'cache built modules to the filesystem',
    }),
    ['cache.type']: Oclif.flags.string({
      default: 'filesystem',
      options: ['filesystem', 'memory', 'false'],
    }),

    clean: Oclif.flags.boolean({
      allowNo: true,
      default: true,
      description: 'clean dist directory before compiling',
    }),

    config: Oclif.flags.string({
      description: 'path to config file',
    }),

    dashboard: Oclif.flags.boolean({
      allowNo: true,
      default: true,
      description: 'enable bud dashboard',
    }),

    devtool: Oclif.flags.string({
      description: 'specify source-map type',
    }),

    html: Oclif.flags.boolean({
      allowNo: true,
      description: 'generate an html template',
    }),

    hash: Oclif.flags.boolean({
      allowNo: true,
      description: 'hash compiled filenames',
    }),

    inject: Oclif.flags.boolean({
      allowNo: true,
      default: true,
      description: 'automatically register & boot extensions',
    }),

    install: Oclif.flags.boolean({
      description: 'ensure peer dependencies are installed',
      default: false,
    }),

    manifest: Oclif.flags.boolean({
      allowNo: true,
      default: true,
      description: 'emit manifest.json',
    }),

    minimize: Oclif.flags.boolean({
      allowNo: true,
      description: 'minimize file size of compiled assets',
    }),

    ['splitChunks']: Oclif.flags.boolean({
      allowNo: true,
      description:
        'create separate chunks for vendor and app code',
    }),

    vendor: Oclif.flags.boolean({
      allowNo: true,
      description:
        'create separate chunks for vendor and app code; alias for splitChunks',
    }),

    runtime: Oclif.flags.boolean({
      allowNo: true,
      description: 'Create a runtime chunk',
    }),

    target: Oclif.flags.string({
      description: 'limit compilation to this compiler',
      multiple: true,
      default: [],
    }),
  }

  public async run() {
    await this.prime(Build)
    await this.build()

    this.app.hooks.on('event.compiler.done', stats => {
      this.notifier.notify(this.app, stats)
      return stats
    })

    await this.app.api.call('run', [])
  }
}
