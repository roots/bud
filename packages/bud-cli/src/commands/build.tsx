import {join} from 'path'
import {yargs} from '@roots/bud-support'

const cwd = process.cwd()

export const aliases: yargs.CommandModule['aliases'] = 'build'

export const describe: yargs.CommandModule['describe'] =
  'Build source into compiled assets.'

export const builder: yargs.CommandModule['builder'] = yargs =>
  yargs
    .option('config', {
      describe: 'Specify a custom configuration file',
      default: 'bud.config.js',
    })
    .option('env', {
      describe: 'Specify build mode',
      choices: ['development', 'production', 'none'],
      default: 'none',
    })
    .option('hot', {
      describe: 'Enable HMR',
      type: 'boolean',
      default: false,
    })
    .option('minify', {
      describe: 'Minify assets',
      type: 'boolean',
      default: false,
    })
    .option('gzip', {
      describe: 'Gzip static assets',
      type: 'boolean',
      default: false,
    })
    .option('hash', {
      describe: 'Hash asset filenames',
      type: 'boolean',
      default: false,
    })
    .option('brotli', {
      describe: 'Apply brotli compression to static assets',
      type: 'boolean',
      default: false,
    })
    .option('runtime', {
      describe: 'Generate runtime chunk',
      type: 'boolean',
      default: false,
    })
    .option('vendor', {
      describe: 'Generate vendor chunk',
      type: 'boolean',
      default: false,
    })
    .option('src', {
      describe: 'Override src directory',
      type: 'string',
    })
    .option('dest', {
      describe: 'Override destination directory',
      type: 'string',
    })
    .option('port', {
      describe: 'Dev port',
      type: 'number',
      default: 3000,
    })
    .option('host', {
      describe: 'Dev host',
      type: 'string',
      default: 'localhost',
    })
    .option('proxy', {
      describe: 'Hostname to proxy',
      type: 'string',
    })
    .option('ci', {
      describe: 'Run build in terminal compatibility mode.',
      type: 'boolean',
      default: false,
    })
    .option('html', {
      describe: 'Use an HTML5 boilerplate',
      type: 'boolean',
    })
    .option('log', {
      describe: 'Specify filepath to log to',
      type: 'string',
    })
    .example('Dev', 'bud build --env development --hot')
    .example(
      'Build',
      'bud build --env production --runtime --vendor --minify --gzip',
    )
    .example(
      'CI',
      'bud build --env production --runtime --vendor --minify --gzip --ci',
    )
    .hide('help')
    .hide('version')
    .showHelpOnFail(true)

/**
 * handler: bud build
 */
export const handler: yargs.CommandModule['handler'] = (args: {
  [config: string]: unknown
}): void => {
  const cfg = args.config ?? 'bud.config.js'
  const cfgPath = join(cwd, cfg as string)

  require(cfgPath)
}
