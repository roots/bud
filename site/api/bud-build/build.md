---
id: bud-build.build
title: build class
sidebar_label: build class
hide_title: true
sidebar: "api"
slug: build
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## Build class

Webpack configuration builder class

Signature:

```typescript
export declare class Build extends Service implements Contract.Interface
```

Extends: Service

Implements: Contract.Interface

## Properties

| Property                                | Modifiers | Type    | Description      |
| --------------------------------------- | --------- | ------- | ---------------- |
| [items](/api/bud-build/build/items)     |           | Items   | Registered items |
| [loaders](/api/bud-build/build/loaders) |           | Loaders |                  |
| [rules](/api/bud-build/build/rules)     |           | Rules   | Registered rules |

## Methods

| Method                                                      | Modifiers | Description                                    |
| ----------------------------------------------------------- | --------- | ---------------------------------------------- |
| [init()](/api/bud-build/build/init)                         |           | Initialize the build rules, loaders, items     |
| [make()](/api/bud-build/build/make)                         |           | Make webpack configuration                     |
| [register()](/api/bud-build/build/register)                 |           | Service register event                         |
| [registered()](/api/bud-build/build/registered)             |           | Service booted event                           |
| [writeFinalConfig()](/api/bud-build/build/writefinalconfig) |           | Write final configuration to storage directory |
