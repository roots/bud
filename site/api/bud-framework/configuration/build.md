---
id: bud-framework.configuration.build
title: build property
sidebar_label: build property
hide_title: true
sidebar: "api"
slug: build
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## Configuration.build property

Initial webpack configuration values

Signature:

```typescript
build: CompilerConfig;
```
