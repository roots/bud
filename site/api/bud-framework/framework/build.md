---
id: bud-framework.framework.build
title: build property
sidebar_label: build property
hide_title: true
sidebar: "api"
slug: build
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## Framework.build property

Build service

Signature:

```typescript
build: Build.Interface;
```
