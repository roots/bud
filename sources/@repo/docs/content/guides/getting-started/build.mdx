---
title: Build
sidebar_label: Build
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'
import CodeBlock from '@theme/CodeBlock'

Use the [`bud build` command](/guides/cli/build) to compile our project.

```bash npm2yarn
npm run bud build
```

## Note for TypeScript users

TypeScript users should use the `ts-bud build` command instead.

```bash npm2yarn
npm run ts-bud build
```
