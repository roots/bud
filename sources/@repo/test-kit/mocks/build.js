import {jest} from '@jest/globals'

const mock = jest.fn().mockImplementation(async () => {
  const mock = {
    logger: {
      log: jest.fn(),
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    config: {},
    label: `build`,
    loaders: {},
    items: {},
    rules: {},
    make: jest.fn(() => ({
      entry: {
        app: [`index.js`],
      },
    })),
    setRule: jest.fn(),
    setLoader: jest.fn(),
    setItem: jest.fn(),
  }

  mock.rules.js = {
    setUse: jest.fn(),
  }

  mock.setLoader = jest.fn(() => mock)
  mock.setItem = jest.fn(() => mock)
  mock.setRule = jest.fn(() => mock)

  return mock
})

export default mock
