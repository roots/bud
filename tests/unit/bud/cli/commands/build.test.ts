import build from '@roots/bud/src/cli/commands/build'

describe('Build Command', () => {
  it('has name: build', () => {
    expect(build.name).toBe('Build')
  })

  it('has description', () => {
    expect(build.description).toBeDefined()
  })

  it(`description is 'compile build assets' `, () => {
    expect(build.description).toBe('compile source assets')
  })

  it('has examples array', () => {
    expect(build.examples).toBeInstanceOf(Array)
  })

  it('has expected flags', () => {
    expect(build.flags).toMatchSnapshot({
      'log.secret': {
        default: [expect.any(String)],
      },
    })
  })
})
