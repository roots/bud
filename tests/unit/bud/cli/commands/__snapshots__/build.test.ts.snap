// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Command has expected flags 1`] = `
Object {
  "cache": Object {
    "allowNo": true,
    "default": true,
    "description": "cache built modules to the filesystem",
    "parse": [Function],
    "type": "boolean",
  },
  "cache.type": Object {
    "default": "filesystem",
    "input": Array [],
    "multiple": false,
    "options": Array [
      "filesystem",
      "memory",
      "false",
    ],
    "parse": [Function],
    "type": "option",
  },
  "clean": Object {
    "allowNo": true,
    "default": true,
    "description": "clean dist directory before compiling",
    "parse": [Function],
    "type": "boolean",
  },
  "config": Object {
    "description": "path to config file",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "dashboard": Object {
    "allowNo": true,
    "default": true,
    "description": "enable bud dashboard",
    "parse": [Function],
    "type": "boolean",
  },
  "devtool": Object {
    "description": "specify source-map type",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "hash": Object {
    "allowNo": true,
    "description": "hash compiled filenames",
    "parse": [Function],
    "type": "boolean",
  },
  "help": Object {
    "allowNo": false,
    "char": "h",
    "description": "Show CLI help.",
    "parse": [Function],
    "type": "boolean",
  },
  "html": Object {
    "allowNo": true,
    "description": "generate an html template",
    "parse": [Function],
    "type": "boolean",
  },
  "inject": Object {
    "allowNo": true,
    "default": true,
    "description": "automatically register & boot extensions",
    "parse": [Function],
    "type": "boolean",
  },
  "install": Object {
    "allowNo": false,
    "default": false,
    "description": "ensure peer dependencies are installed",
    "parse": [Function],
    "type": "boolean",
  },
  "location.dist": Object {
    "description": "distribution directory",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "location.modules": Object {
    "description": "public path",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "location.project": Object {
    "description": "repo root path",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "location.publicPath": Object {
    "description": "public path",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "location.src": Object {
    "description": "source directory",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "location.storage": Object {
    "description": "storage directory",
    "input": Array [],
    "multiple": false,
    "parse": [Function],
    "type": "option",
  },
  "log": Object {
    "allowNo": true,
    "default": true,
    "description": "log to console",
    "parse": [Function],
    "type": "boolean",
  },
  "log.level": Object {
    "default": "vvv",
    "description": "set log verbosity. \`v\` is error level. \`vv\` is warning level. \`vvv\` is log level. \`vvvv\` is debug level.",
    "input": Array [],
    "multiple": false,
    "options": Array [
      "v",
      "vv",
      "vvv",
      "vvvv",
    ],
    "parse": [Function],
    "type": "option",
  },
  "log.min": Object {
    "allowNo": true,
    "default": true,
    "description": "remove formatting from logged objects",
    "parse": [Function],
    "type": "boolean",
  },
  "log.papertrail": Object {
    "allowNo": true,
    "default": true,
    "description": "preserve logger output",
    "parse": [Function],
    "type": "boolean",
  },
  "log.secret": Object {
    "default": Array [
      Any<String>,
    ],
    "description": "hide matching strings from logging output",
    "input": Array [],
    "multiple": true,
    "parse": [Function],
    "type": "option",
  },
  "manifest": Object {
    "allowNo": true,
    "default": true,
    "description": "emit manifest.json",
    "parse": [Function],
    "type": "boolean",
  },
  "minimize": Object {
    "allowNo": true,
    "description": "minimize file size of compiled assets",
    "parse": [Function],
    "type": "boolean",
  },
  "mode": Object {
    "default": "production",
    "description": "compiler mode",
    "hidden": true,
    "input": Array [],
    "multiple": false,
    "options": Array [
      "development",
      "production",
    ],
    "parse": [Function],
    "type": "option",
  },
  "runtime": Object {
    "allowNo": true,
    "description": "Create a runtime chunk",
    "parse": [Function],
    "type": "boolean",
  },
  "splitChunks": Object {
    "allowNo": true,
    "description": "create separate chunks for vendor and app code",
    "parse": [Function],
    "type": "boolean",
  },
  "target": Object {
    "default": Array [],
    "description": "limit compilation to this compiler",
    "input": Array [],
    "multiple": true,
    "parse": [Function],
    "type": "option",
  },
  "vendor": Object {
    "allowNo": true,
    "description": "create separate chunks for vendor and app code; alias for splitChunks",
    "parse": [Function],
    "type": "boolean",
  },
  "version": Object {
    "allowNo": false,
    "description": "Show CLI version.",
    "parse": [Function],
    "type": "boolean",
  },
}
`;
